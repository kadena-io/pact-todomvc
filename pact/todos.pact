;;
;; todos smart contract
;;

;; admin keyset definition
(define-keyset 'todo-admin-keyset
  (read-keyset "todo-admin-keyset"))

;; todos module, administered by keyset
(module todos 'todo-admin-keyset
  " Smart contract module for TODO-MVC pact app.   \
  \ Tables:                                        \
  \ todo -- holds todo entries"

  ;; todo schema and table
  (defschema todo
    "Row type for todos."
     title:string
     completed:bool
     deleted:bool
     )

  (deftable todo-table:{todo})

  ;;
  ;; API functions
  ;;

  (defun new-todo (id title)
    "Create new todo with ENTRY and DATE."
    (insert todo-table id {
      "title": title,
      "completed": false,
      "deleted": false })
  )

  (defun toggle-todo-status (id:string)
    "Toggle completed status flag for todo at ID."
    (with-read todo-table id {
      "completed":= state
      }
      (update todo-table id {
        "completed": (not state) })))

  (defun edit-todo (id:string title)
    "Update todo ENTRY at ID."
    (update todo-table id {
      "title": title }))

  (defun delete-todo (id:string)
    "Delete todo title at ID (by setting deleted flag)."
    (update todo-table id {
      "deleted": true }))

  (defun read-todo:object (id:string)
    "Read a single todo"
    (+ {'id: id} (read todo-table id)))

  (defun read-todos:[object{todo}] ()
    "Read all todos."
    (map (read-todo) (keys todo-table)))
)

(create-table todo-table)
;done
