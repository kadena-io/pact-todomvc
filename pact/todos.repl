(env-keys ["todoadmin"])
(env-data { "todo-admin-keyset":
  { "keys": ["todoadmin"], "pred": "keys-all" } })

(begin-tx)
(load "todos.pact")
(commit-tx)

(typecheck "todos")

(use todos)

;;Create a new todo
(new-todo "1" "wash")

;;Test if todo was created
(expect "Returns the todo object with id-1" {"completed": false, "deleted": false, "id": "1", "title": "wash"} (read-todo "1"))

;;Fail if todo to be edited does not exist.
(expect-failure "No todo exists with id of 2." (edit-todo "2" "clean"))
