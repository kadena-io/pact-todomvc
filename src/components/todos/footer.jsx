import * as React from "react";

export class TodoFooter extends React.PureComponent {
  render() {
    var pluralize = function(count, word) {
      return count === 1 ? word : word + "s";
    };
    var activeTodoWord = pluralize(this.props.count, "item");
    var clearButton = null;

    if (this.props.completedCount > 0) {
      clearButton = (
        <button
          className="clear-completed"
          onClick={this.props.onClearCompleted}
        >
          Clear completed
        </button>
      );
    }

    var nowShowing = this.props.nowShowing;
    return (
      <footer className="footer">
        <span className="todo-count">
          <strong>{this.props.count}</strong> {activeTodoWord} left
        </span>
        <ul className="filters">
          <li>
            <a href="#/" className={{ selected: nowShowing === "all" }}>
              All
            </a>
          </li>{" "}
          <li>
            <a
              href="#/active"
              className={{ selected: nowShowing === "active" }}
            >
              Active
            </a>
          </li>{" "}
          <li>
            <a
              href="#/completed"
              className={{ selected: nowShowing === "completed" }}
            >
              Completed
            </a>
          </li>
        </ul>
        {clearButton}
      </footer>
    );
  }
}
