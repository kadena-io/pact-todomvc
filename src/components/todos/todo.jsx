import * as React from "react";

var ESCAPE_KEY = 27;
var ENTER_KEY = 13;

export class TodoItem extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = { editText: this.props.todo.title };
  }

  handleSubmit(event) {
    var val = this.state.editText.trim();
    if (val) {
      this.props.onSave(val);
      this.setState({ editText: val });
    } else {
      this.props.onDestroy();
    }
  }

  handleEdit() {
    this.props.onEdit();
    this.setState({ editText: this.props.todo.title });
  }

  handleKeyDown(event) {
    if (event.which === ESCAPE_KEY) {
      this.setState({ editText: this.props.todo.title });
      this.props.onCancel(event);
    } else if (event.which === ENTER_KEY) {
      this.handleSubmit(event);
    }
  }

  handleChange(event) {
    if (this.props.editing) {
      this.setState({ editText: event.target.value });
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      nextProps.todo !== this.props.todo ||
      nextProps.editing !== this.props.editing ||
      nextState.editText !== this.state.editText
    );
  }
  componentDidUpdate(prevProps) {
    if (!prevProps.editing && this.props.editing) {
      var node = React.findDOMNode(this.refs.editField);
      node.focus();
      node.setSelectionRange(node.value.length, node.value.length);
    }
  }

  render() {
    return (
      <li
        className={{
          completed: this.props.todo.completed,
          editing: this.props.editing
        }}
      >
        <div className="view">
          <input
            className="toggle"
            type="checkbox"
            checked={this.props.todo.completed}
            onChange={this.props.onToggle}
          />
          <label onDoubleClick={this.handleEdit}>{this.props.todo.title}</label>
          <button className="destroy" onClick={this.props.onDestroy} />
        </div>
        <input
          ref="editField"
          className="edit"
          value={this.props.editText}
          onBlur={this.handleSubmit}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
        />
      </li>
    );
  }
}
