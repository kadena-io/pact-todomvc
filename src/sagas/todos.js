import { call, put } from 'redux-saga/effects';

import sendPactCommand from '../utils/send-pact-command';

export const FETCH_TODOS = 'pact-todomvc/todos/FETCH_TODOS';
export const FETCH_TODOS_REQUEST = 'pact-todomvc/todos/FETCH_TODOS_REQUEST';
export const FETCH_TODOS_SUCCEEDED = 'pact-todomvc/todos/FETCH_TODOS_SUCCEEDED';
export const FETCH_TODOS_FAILED = 'pact-todomvc/todos/FETCH_TODOS_FAILED';

export const SAVE_NEW_TODO = 'pact-todomvc/todos/SAVE_NEW_TODO';
export const SAVE_NEW_TODO_REQUEST = 'pact-todomvc/todos/SAVE_NEW_TODO_REQUEST';
export const SAVE_NEW_TODO_SUCCEEDED = 'pact-todomvc/todos/SAVE_NEW_TODO_SUCCEEDED';
export const SAVE_NEW_TODO_FAILED = 'pact-todomvc/todos/SAVE_NEW_TODO_FAILED';

export const REMOVE_TODO = 'pact-todomvc/todos/REMOVE_TODO';
export const REMOVE_TODO_REQUEST = 'pact-todomvc/todos/REMOVE_TODO_REQUEST';
export const REMOVE_TODO_SUCCEEDED = 'pact-todomvc/todos/REMOVE_TODO_SUCCEEDED';
export const REMOVE_TODO_FAILED = 'pact-todomvc/todos/REMOVE_TODO_FAILED';

export const UPDATE_TODO = 'pact-todomvc/todos/UPDATE_TODO';
export const UPDATE_TODO_REQUEST = 'pact-todomvc/todos/UPDATE_TODO_REQUEST';
export const UPDATE_TODO_SUCCEEDED = 'pact-todomvc/todos/UPDATE_TODO_SUCCEEDED';
export const UPDATE_TODO_FAILED = 'pact-todomvc/todos/UPDATE_TODO_FAILED';

export const TOGGLE_TODO_STATE = 'pact-todomvc/todos/TOGGLE_TODO_STATE';
export const TOGGLE_TODO_STATE_REQUEST = 'pact-todomvc/todos/TOGGLE_TODO_STATE_REQUEST';
export const TOGGLE_TODO_STATE_SUCCEEDED = 'pact-todomvc/todos/TOGGLE_TODO_STATE_SUCCEEDED';
export const TOGGLE_TODO_STATE_FAILED = 'pact-todomvc/todos/TOGGLE_TODO_STATE_FAILED';

export const CHANGE_EDIT_STATUS = 'pact-todomvc/todos/CHANGE_EDIT_STATUS';
export const CHANGE_TODO_ENTRY = 'pact-todomvc/todos/CHANGE_TODO_ENTRY';
export const UPDATE_NEW_TODO_FIELD = 'pact-todomvc/todos/UPDATE_NEW_TODO_FIELD';

const initialState = {
  todosIsLoading: false,
  todosError: null,
  todos: [],
  newTodo: '',
  editStatus: false,
  editedTodo: null,
};

let workingTodo;
export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case FETCH_TODOS_REQUEST:
      return { ...state, todosIsLoading: true };

    case FETCH_TODOS_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: action.todos
          .sort((a, b) => a.id - b.id)
          .map(todo => ({ ...todo, editStatus: false })),
        todosIsLoading: false,
      };

    case FETCH_TODOS_FAILED:
      return {
        ...state,
        todosError: action.error,
        todos: [],
        todosIsLoading: false,
      };

    case SAVE_NEW_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case SAVE_NEW_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: [...state.todos, action.newTodo].sort((a, b) => a.id - b.id),
        todosIsLoading: false,
        newTodo: '',
      };

    case SAVE_NEW_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case REMOVE_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case REMOVE_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: state.todos.filter(todo => todo.id !== action.id),
        todosIsLoading: false,
        newTodo: '',
      };

    case REMOVE_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case UPDATE_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case UPDATE_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: [
          ...state.todos.filter(todo => todo.id !== action.todo.id),
          { ...action.todo, editStatus: false },
        ].sort((a, b) => a.id - b.id),
        todosIsLoading: false,
      };

    case UPDATE_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case TOGGLE_TODO_STATE_REQUEST:
      return { ...state, todosIsLoading: true };

    case TOGGLE_TODO_STATE_SUCCEEDED:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.state = action.state;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
        todosError: null,
        todosIsLoading: false,
      };

    case TOGGLE_TODO_STATE_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case CHANGE_TODO_ENTRY:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.entry = action.entry;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
      };

    case CHANGE_EDIT_STATUS:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.editStatus = !workingTodo.editStatus;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
      };

    case UPDATE_NEW_TODO_FIELD:
      return { ...state, newTodo: action.newTodo };

    default:
      return state;
  }
}

export function updateNewTodoField(newTodo) {
  return { type: UPDATE_NEW_TODO_FIELD, newTodo };
}

export function fetchTodos() {
  return { type: FETCH_TODOS };
}

export function saveNewTodo(entry, date) {
  return { type: SAVE_NEW_TODO, entry, date };
}

export function changeEntry(id, entry, date) {
  return {
    type: CHANGE_TODO_ENTRY,
    id,
    entry,
    date,
  };
}

export function updateTodo(todo) {
  return { type: UPDATE_TODO, todo };
}

export function toggleState(id, state) {
  return { type: TOGGLE_TODO_STATE, id, state };
}

export function removeTodo(id) {
  return { type: REMOVE_TODO, id };
}

export function changeEditStatus(id) {
  return { type: CHANGE_EDIT_STATUS, id };
}

export function* fetchTodosSaga() {
  yield put({ type: FETCH_TODOS_REQUEST });
  try {
    const todos = yield call(sendPactCommand, '(todos.read-todos)');
    yield put({ type: FETCH_TODOS_SUCCEEDED, todos });
  } catch (error) {
    yield put({ type: FETCH_TODOS_FAILED, error });
  }
}

export function* toggleTodoStateSaga({ id, state }) {
  yield put({ type: TOGGLE_TODO_STATE_REQUEST });
  try {
    yield call(sendPactCommand, `(todos.toggle-todo-status ${id})`);
    yield put({ type: TOGGLE_TODO_STATE_SUCCEEDED, id, state });
  } catch (error) {
    yield put({ type: TOGGLE_TODO_STATE_FAILED, error });
  }
}

export function* saveNewTodoSaga({ entry, date }) {
  yield put({ type: SAVE_NEW_TODO_REQUEST });
  try {
    const newTodo = yield call(
      sendPactCommand,
      `(todos.new-todo ${JSON.stringify(entry)} ${JSON.stringify(date)})`
    );
    yield put({ type: SAVE_NEW_TODO_SUCCEEDED, newTodo });
  } catch (error) {
    yield put({ type: SAVE_NEW_TODO_FAILED, error });
  }
}

export function* removeTodoSaga({ id }) {
  yield put({ type: REMOVE_TODO_REQUEST, id });
  try {
    yield call(sendPactCommand, `(todos.delete-todo ${id})`);
    yield put({ type: REMOVE_TODO_SUCCEEDED, id });
  } catch (error) {
    yield put({ type: REMOVE_TODO_FAILED, error });
  }
}

export function* updateTodoSaga({ todo }) {
  yield put({ type: UPDATE_TODO_REQUEST, todo });
  try {
    yield call(
      sendPactCommand,
      `(todos.edit-todo ${todo.id} ${JSON.stringify(todo.entry)}  ${JSON.stringify(todo.date)})`
    );
    yield put({ type: UPDATE_TODO_SUCCEEDED, todo });
  } catch (error) {
    yield put({ type: UPDATE_TODO_FAILED, error });
  }
}
