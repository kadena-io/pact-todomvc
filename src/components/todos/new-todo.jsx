import * as React from 'react';
import './todo.scss';

export const NewTodo = ({ saveNewTodo }) => {
  let entry='';
  let date;

  const handleDate = e => {
    e.preventDefault();
    date = e.currentTarget.value;
  };

  const handleEntry = e => {
    e.preventDefault();
    entry = e.currentTarget.value;
  };

  const entryKeyDown = e => {
    if (e.keyCode === 13) {
      let newDate = new Date();
      date = date || newDate.toISOString().slice(0,10);
      console.log(entry, date)
      saveNewTodo(entry, date);
      entry='';
    }
  };

  return (
    <div className="todo new">
      <div className="entry">
        <label>New Task</label>
        <input id= "new-todo" type="text" className="empty" onChange={handleEntry} onKeyDown={entryKeyDown} placeholder="New Todoâ€¦" />
      </div>
      <div className="date">
        <label>Due Date</label>
        <input id="due-date" type="date" onChange={handleDate} onKeyDown={entryKeyDown} placeholder="Enter Due Date..." />
      </div>
    </div>
  );
};
