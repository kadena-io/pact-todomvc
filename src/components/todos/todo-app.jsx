import * as React from "react";

import { TodoItem } from "./todo.jsx";
import { TodoFooter } from "./footer.jsx";
import uuidv4 from "uuid/v4";
import Pact from "pact-lang-api";

const ENTER_KEY = 13;
const KP = Pact.crypto.genKeyPair();
const API_HOST = "http://localhost:9001";
const ALL_TODOS = "all";
const ACTIVE_TODOS = "active";
const COMPLETED_TODOS = "completed";

export class TodoApp extends React.PureComponent {
  constructor() {
    super();
    this.state = {
      todos: [],
      onChanges: [],
      nowShowing: ALL_TODOS,
      editing: null,
      newTodo: ""
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleNewTodoKeyDown = this.handleNewTodoKeyDown.bind(this);
    this.toggle = this.toggle.bind(this);
    this.toggleAll = this.toggleAll.bind(this);
    this.destroy = this.destroy.bind(this);
    this.edit = this.edit.bind(this);
    this.cancel = this.cancel.bind(this);
    this.clearCompleted = this.clearCompleted.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.showActive = this.showActive.bind(this);
    this.showCompleted = this.showCompleted.bind(this);
    this.showAll =this.showAll.bind(this);
  }

  componentDidMount() {
    this.getTodos();
  }

  getTodos() {
    const cmdObj = {
      pactCode: `(todos.read-todos)`,
      keyPairs: KP
    };
    return Pact.fetch.local(cmdObj, API_HOST)
      .then(res => res.data)
      .then(todos =>{
        const notDeleted = todos.filter(todo => {
          return todo.deleted===false
        })
        this.setState({ todos: notDeleted });
      })
      .catch(e => e);
  }

  handleChange(event) {
    this.setState({ newTodo: event.target.value });
  }

  handleNewTodoKeyDown(event) {
    if (event.keyCode !== ENTER_KEY) {
      return;
    }

    event.preventDefault();

    var val = this.state.newTodo.trim();

    if (val) {
      this.add(val);
      this.setState({ newTodo: "" });
    }
  }

  add(title) {
    const uuid = uuidv4();
    const cmdObj = {
      pactCode: `(todos.new-todo ${JSON.stringify(uuid)} ${JSON.stringify(title)})`,
      keyPairs: KP
    };

    Pact.fetch.send(cmdObj, API_HOST)
    this.getTodos();
  }

  toggle(todo) {
    const cmdObj = {
      pactCode: `(todos.toggle-todo-status
                 ${JSON.stringify(todo.id)})`,
      keyPairs: KP
    };
    Pact.fetch.send(cmdObj, API_HOST);
    this.getTodos();
  }

  toggleAll(){
    const activeTodos = this.state.todos.filter(todo => !todo.completed)
    const cmds = activeTodos.map(todo => {
      return {
        pactCode:`(todos.toggle-todo-status
                   ${JSON.stringify(todo.id)})`,
        keyPairs: KP
    }});
    Pact.fetch.send(cmds, API_HOST)
    this.getTodos();
  }

  destroy(todo) {
    const cmdObj = {
      pactCode: `(todos.delete-todo
                 ${JSON.stringify(todo.id)})`,
      keyPairs: KP
    };
    Pact.fetch.send(cmdObj, API_HOST);
    this.getTodos();
  }

  edit(todo, text) {
    const cmdObj = {
      pactCode: `(todos.edit-todo
                ${JSON.stringify(todo.id)})
                ${JSON.stringify(text)})`,
      keyPairs: KP
    };
    Pact.fetch.send(cmdObj, API_HOST);
    this.setState({ editing: null });
    this.getTodos();
  }

  cancel() {
    this.setState({ editing: null });
  }

  clearCompleted() {
    const completedTodos = this.state.todos.filter(todo => todo.completed)
    const cmds = completedTodos.map(todo => {
      return {
        pactCode: `(todos.delete-todo
                   ${JSON.stringify(todo.id)})`,
        keyPairs: KP
    }});
    Pact.fetch.send(cmds, API_HOST)
    this.getTodos();
  }

  showActive(){
    this.setState({nowShowing: ACTIVE_TODOS})
  }

  showCompleted(){
    this.setState({nowShowing: COMPLETED_TODOS})
  }

  showAll(){
    this.setState({nowShowing: ALL_TODOS})
  }

  handleSubmit(e) {
    var val = this.state.editText.trim();
    if (val) {
      this.save(val);
      this.setState({
        editText: val
      });
    } else {
      this.destroy();
      this.getTodos();
    }
  }

  render() {
    var footer;
    var main;
    var todos = this.state.todos;
    var shownTodos = todos.filter(function(todo) {
      switch (this.state.nowShowing) {
        case ACTIVE_TODOS:
          return !todo.completed;
        case COMPLETED_TODOS:
          return todo.completed;
        default:
          return true;
      }
    }, this);

    var todoItems = shownTodos.map(function(todo) {
      return (
        <TodoItem
          key={todo.id}
          todo={todo}
          onToggle={this.toggle.bind(this, todo)}
          onDestroy={this.destroy.bind(this, todo)}
          onEdit={this.edit.bind(this, todo)}
          editing={this.state.editing === todo.id}
          onCancel={this.cancel}
        />
      );
    }, this);

    var activeTodoCount = todos.reduce(function(accum, todo) {
      return todo.completed ? accum : accum + 1;
    }, 0);

    var completedCount = todos.length - activeTodoCount;

    if (activeTodoCount || completedCount) {
      footer = (
        <TodoFooter
          count={activeTodoCount}
          completedCount={completedCount}
          nowShowing={this.state.nowShowing}
          onClearCompleted={this.clearCompleted}
          showActive={this.showActive}
          showCompleted = {this.showCompleted}
          showAll = {this.showAll}
        />
      );
    }

    if (todos.length) {
      main = (
        <section className="main">
          <input
            id="toggle-all"
            className="toggle-all"
            type="checkbox"
            onChange={this.toggleAll}
            checked={activeTodoCount === 0}
          />
          <label htmlFor="toggle-all" />
          <ul className="todo-list">{todoItems}</ul>
        </section>
      );
    }

    return (
      <div>
        <header className="header">
          <h1>todos</h1>
          <input
            className="new-todo"
            placeholder="What needs to be done?"
            value={this.state.newTodo}
            onKeyDown={this.handleNewTodoKeyDown}
            onChange={this.handleChange}
            autoFocus={true}
          />
        </header>
        {main}
        {footer}
      </div>
    );
  }
}
