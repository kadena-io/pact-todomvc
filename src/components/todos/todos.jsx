import * as React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Loader from 'react-loader-spinner';

import {
  fetchTodos,
  saveNewTodo,
  removeTodo,
  changeEntry,
  updateTodo,
  toggleState,
  updateNewTodoField,
} from '../../sagas/todos';

import { NewTodo } from './new-todo';
import { Todo } from './todo';

export class TodosComponent extends React.PureComponent {
  constructor() {
    super();
    this.fetchTodos = this.fetchTodos.bind(this);
    this.onRemoveTodo = this.onRemoveTodo.bind(this);
    this.onUpdateTodo = this.onUpdateTodo.bind(this);
    this.onToggleState = this.onToggleState.bind(this);
    this.onChangeEntry = this.onChangeEntry.bind(this);
    this.saveNewTodo = this.saveNewTodo.bind(this);
  }

  componentDidMount() {
    this.fetchTodos();
  }

  fetchTodos() {
    this.props.fetchTodos();
  }

  onRemoveTodo(id) {
    this.props.removeTodo(id);
  }

  onUpdateTodo(todo) {
    this.props.updateTodo(todo);
  }

  onToggleState(id, state) {
    this.props.toggleState(id, state);
  }

  onChangeEntry(id, entry) {
    this.props.changeEntry(id, entry);
  }

  saveNewTodo(entry) {
    this.props.saveNewTodo(entry);
  }

  render() {
    const { todos, todosIsLoading, todosError } = this.props;
    let dom;
    if (todosIsLoading) {
      dom = (
        <div className="loading">
          <Loader type="TailSpin" color="#3cf" />
        </div>
      );
    } else if (todosError !== null) {
      dom = <div className="error">Error loading todos</div>;
    } else {
      dom = (
        <div>
          <div className="show-completed">
            <input type="checkbox" /> Show Completed Items
          </div>
          <div className="todos-list">
            <NewTodo saveNewTodo={this.saveNewTodo} />
            {todos.map((todo, i) => (
              <Todo
                key={i}
                {...todo}
                onRemove={this.onRemoveTodo}
                onChangeEntry={this.onChangeEntry}
                onUpdate={this.onUpdateTodo}
                onToggleState={this.onToggleState}
              />
            ))}
          </div>
        </div>
      );
    }

    return (
      <div className="todos">
        <h1>Todos</h1>
        {dom}
      </div>
    );
  }
}

const mapStateToProps = state => {
  /*
  todosIsLoading: false,
  todosError: null,
  todos: [],
  newTodo: '',
  editedTodo: null,
  */
  return { ...state.todos, todosIsLoading: true };
};

const mapDispatchToProps = dispatch => {
  return {
    fetchTodos: () => {
      dispatch(fetchTodos());
    },
    saveNewTodo: newTodo => {
      dispatch(saveNewTodo(newTodo));
    },
    removeTodo: id => {
      dispatch(removeTodo(id));
    },
    updateTodo: todo => {
      dispatch(updateTodo(todo));
    },
    toggleState: (id, state) => {
      dispatch(toggleState(id, state));
    },
    changeEntry: (id, entry) => {
      dispatch(changeEntry(id, entry));
    },
    updateNewTodoField: entry => {
      dispatch(updateNewTodoField(entry));
    },
  };
};

export const Todos = connect(
  mapStateToProps,
  mapDispatchToProps
)(TodosComponent);
