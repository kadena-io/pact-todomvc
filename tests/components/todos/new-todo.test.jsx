import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { NewTodo } from '../../../src/components/todos/new-todo';

const newTodoProps = { saveNewTodo: jest.fn() };

describe('New Todo Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render a new todo', () => {
    const component = renderer.create(<NewTodo {...newTodoProps} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should only call saveNewTodo() on text field if keyDown keyCode is 13', () => {
    const component = shallow(<NewTodo {...newTodoProps} />);
    const input = component.find('.entry input');
    input.value = 'Something Else';

    input.simulate('keyDown', { keyCode: 1, currentTarget: input });
    expect(newTodoProps.saveNewTodo).not.toHaveBeenCalled();
    expect(input.value).toEqual(input.value);

    input.simulate('keyDown', { keyCode: 13, currentTarget: input });
    expect(newTodoProps.saveNewTodo).toHaveBeenCalled();
    expect(input.value).toEqual('');
  });
});
