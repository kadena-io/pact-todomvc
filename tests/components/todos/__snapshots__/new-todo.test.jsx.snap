// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Todo Component should render a new todo 1`] = `
<div
  className="todo new"
>
  <div
    className="entry"
  >
    <input
      className="empty"
      onKeyDown={[Function]}
      placeholder="New Todoâ€¦"
      type="text"
    />
  </div>
</div>
`;
